File: solutions/853.car-fleet.java
/*
 * @lc app=leetcode id=853 lang=java
 *
 * [853] Car Fleet
 */

// @lc code=start

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int carFleet(int target, int[] position, int[] speed) {
        Map<Integer, Integer> posMap = new HashMap<>();
        for (int i = 0; i < position.length; i++) {
            posMap.merge(position[i], i, (prev, curr) -> curr);
        }

        int[] sorted = Arrays.copyOf(position, position.length);
        Arrays.sort(sorted);

        double prevCarTime = -1;
        int fleets = 0;
        for (int i = sorted.length - 1; i >= 0; i--) {
            int pos = sorted[i];
            double carTime = (double) (target - pos) / speed[posMap.get(pos)];
            if (prevCarTime < carTime) {
                fleets++;
                prevCarTime = carTime;
            }
        }

        return fleets;
    }
}
// @lc code=end

File: solutions/853.car-fleet.py
#
# @lc app=leetcode id=853 lang=python3
#
# [853] Car Fleet
#

# @lc code=start
from typing import List


class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        cars = sorted(zip(position, speed), reverse=True)
        fleets = 0
        lead_car_time = -1

        for pos, spd in cars:
            time_to_reach_target = (target - pos) / spd

            if time_to_reach_target > lead_car_time:
                fleets += 1
                lead_car_time = time_to_reach_target

        return fleets

# @lc code=end

